OpenZFS 7968 - multi-threaded spa_sync

ahrens opened the pull request:
https://github.com/zfsonlinux/zfs/issues/4807

`spa_sync()` iterates over all the dirty dnodes and processes each of them by calling `dnode_sync()`.  If there are many dirty dnodes (e.g. because we created or removed a lot of files), the single thread of `spa_sync()` calling `dnode_sync()` can become a bottleneck.  Additionally, if many dnodes are dirtied concurrently in open context (e.g. due to concurrent file creation), the `os_lock` will experience lock contention via `dnode_setdirty()`.

Work sponsored by Intel Corp.

### Description
The solution is to track dirty dnodes on a `multilist_t`, and for `spa_sync()` to use separate threads to process each of the sublists in the multilist.

### Motivation and Context
Improve performance when many files are created/deleted/modified.
https://github.com/zfsonlinux/zfs/issues/4807

### How Has This Been Tested?
Equivalent changes on illumos have passed ztest and the zfs test suite (https://github.com/openzfs/openzfs/pull/138).
Minimal performance testing on Linux at this point.

On the concurrent file creation microbenchmark, the performance improvement from `dnode_setdirty()` is up to 7%.  Additionally, the wall clock time spent in `spa_sync()` is reduced to 15%-40% of the single-threaded case.

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Performance enhancement (non-breaking change which improves efficiency)
- [x] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the ZFS on Linux code style requirements.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [ ] Change has been approved by a ZFS on Linux member.

behlendorf commented on module/zfs/dmu_objset.c:
```diff
@@ -499,10 +524,10 @@ dmu_objset_open_impl(spa_t *spa, dsl_dataset_t *ds, blkptr_t *bp,
-		    offsetof(dnode_t, dn_dirty_link[i]));
+		multilist_create(&os->os_dirty_dnodes[i], sizeof (dnode_t),
+		    offsetof(dnode_t, dn_dirty_link[i]),
+		    zfs_arc_num_sublists_per_state,
```
It seems plausible that we'd want to tune this independently of `zfs_arc_num_sublists_per_state`.  I'm guessing you used this because it defaults to number of cpus?

ahrens replied:
Yes, this was mostly out of convenience.  In theory we might want it to be different, but in practice I would be surprised if that ever happened.  I don't mind making it a different variable if you want, but another option would be to rename/move the existing variable to be more obviously generic (e.g. `zfs_multilist_num_sublists`)

behlendorf replied:
Me too.  Going with `zfs_multilist_num_sublists` seems like a good way to go since it'll give us a generic option for this kind of thing.

ahrens replied:
I opened https://github.com/zfsonlinux/zfs/pull/5764 to do this.  I'll rebase this on top of that.

behlendorf commented on module/zfs/dmu_objset.c:
```diff
@@ -342,6 +342,31 @@ dmu_objset_byteswap(void *buf, size_t size)
+	crc ^= (osv>>14) ^ (obj>>24);
+
+	return (crc);
+}
```
A short comment describing why the hash is composed this way would be nice.

ahrens replied:
OK

behlendorf commented on module/zfs/dmu_objset.c:
```diff
@@ -1290,12 +1341,24 @@ dmu_objset_sync(objset_t *os, zio_t *pio, dmu_tx_t *tx)
-	dmu_objset_sync_dnodes(&os->os_dirty_dnodes[txgoff], newlist, tx);
+	for (int i = 0;
+	    i < multilist_get_num_sublists(&os->os_dirty_dnodes[txgoff]); i++) {
+		sync_dnodes_arg_t *sda = kmem_alloc(sizeof (*sda), KM_SLEEP);
```
It would be nice to use a kmem cache for these small short lived allocations.  It may not make a significant performance impact but as a fringe benefit it provides us additional visibility in to this machinery.

behlendorf commented on module/zfs/dmu_objset.c:
```diff
@@ -1388,17 +1450,26 @@ do_userquota_cacheflush(objset_t *os, userquota_cache_t *cache, dmu_tx_t *tx)
+		mutex_enter(&os->os_userused_lock);
 		VERIFY0(zap_increment(os, DMU_GROUPUSED_OBJECT,
 		    uqn->uqn_id, uqn->uqn_delta, tx));
+		mutex_exit(&os->os_userused_lock);
```
It's good thing these trees are expected to be small.  It might be worth breaking these in to cacheline aligned `os_userused_lock` and `os_groupused_lock` locks if there's significant contention.

ahrens replied:
The number of modified groups is likely to be much less than the number of modified users, so in the common case I don't think that would help much.  Note that the lock is per-objset, but we can concurrently call this for every objset in the pool, so if there are several modified objsets, we will still have decent concurrency here. (But as you say, in the common case this shouldn't be much of a problem - and this change isn't making things any worse.)

behlendorf commented on module/zfs/dnode.c:
```diff
@@ -1441,11 +1441,7 @@ dnode_setdirty(dnode_t *dn, dmu_tx_t *tx)
-	} else {
-		list_insert_tail(&os->os_dirty_dnodes[txg&TXG_MASK], dn);
-	}
+	multilist_insert(&os->os_dirty_dnodes[txg&TXG_MASK], dn);
```
[cstyle] It would be nice to add the white space back around the `&` in `[txg&TXG_MASK]`.

ahrens replied:
Good idea, will do.

behlendorf commented on include/sys/dmu_objset.h:
```diff
@@ -124,11 +124,13 @@ struct objset {
 	/* Protected by os_lock */
 	kmutex_t os_lock;
-	list_t os_dirty_dnodes[TXG_SIZE];
-	list_t os_free_dnodes[TXG_SIZE];
```
I'm happy to see `os_free_dnodes` go.  I've found it confusing in the past since care is taken to maintain them for no apparent purpose.

ahrens commented on usr/src/test/zfs-tests/tests/functional/snapshot/deadlist_lock.ksh:
```diff
@@ -0,0 +1,104 @@
+function cleanup
+{
+	log_must kill -9 $DLOOP_PID $DTRACE_PID
+	log_must zfs destroy -fR $TESTPOOL/$TESTFS/$DLDS
```
We now have a different method for finding the utilities, we modify the `$PATH` to contain only a directory that has symlinks to the utilities.  This was changed by the following illumos commit:
```
commit 1d32ba663e202c24a5a1f2e5aef83fffb447cb7f
Author: John Wren Kennedy <john.kennedy@delphix.com>
Date:   Fri Nov 11 09:57:19 2016 -0800

    7290 ZFS test suite needs to control what utilities it can run
```

behlendorf replied:
Ugh.  We could probably do something similar, but it feels like a lot of pain for not much gain.

ahrens commented:
@behlendorf 
> The deadlist changes look like they were just brought in as a dependency, correct?

That's correct, and I can break those out into a separate PR (which will also help by segregating the test suite questions into a separate PR).

bzzz77 commented:
is there a way to download the patch in the browser?

behlendorf commented:
@ahrens if you can break DLPX-47082 in to a separate PR it'll be easier for us to address the test suite issues and get merged.  There's no harm in keep the patch in this PR too until we get the first one merged.

It looks like there are a few [style](http://build.zfsonlinux.org/builders/Amazon%202015.09%20x86_64%20%28STYLE%29/builds/2616/steps/shell_2/logs/stdio) issues caught by cstyle which need to be fixed.  We should now have all the same style checking enabled as OpenZFS.

@bzzz77 you can view each commit independently in the browser by selecting the hash.  But unfortunately, I don't know of a good way to see them all squashed.

behlendorf commented:
Rebasing on the latest master which includes 8eecd4a2de217b3f7a40982dcf7765bc0c4104ba should resolve the reservation test case failures.  It's a fix to the test suite which should go back to OpenZFS.

kernelOfTruth commented:
@bzzz77 add .diff to the address of the pully request:

https://github.com/zfsonlinux/zfs/pull/5752.diff

then becomes

https://patch-diff.githubusercontent.com/raw/zfsonlinux/zfs/pull/5752.diff

bzzz77 commented:
@behlendorf  head-to-head as suggested above:

> clean       5752      5748
> -----------------------
> 28186     28577    28801
> 52995     53185    54996
> 101170     97052    104859
> 152050    150715   176329
> 137160    135298   165544
> 135367    132683   161457
> 135482    137292   162365
> 138157     139322   161378

behlendorf commented:
@bzzz77 surprising!  It seems like we're still seeing a lot of lock contention with the default turnings in #5752.  Could you try increasing `zfs_arc_num_sublists_per_state` to 16 or even larger to increase the parallelism.  Perhaps we're getting a lot of hash collisions.

bzzz77 commented:
@behlendorf I think this is because of os_lock in dnode_setdirty() which I was trying to fix in my patch. multilist itself doesn't fix this bottleneck AFAIU.
I can re-benchmark as you suggested.

ahrens commented:
@bzzz77 @behlendorf 
> this is because of os_lock in dnode_setdirty() which I was trying to fix in my patch. multilist itself doesn't fix this bottleneck

That's right.  I didn't read your patch closely enough, sorry.  Let me see if I can address that in this patch, by making dnode_setdirty() use the sublist's lock rather than os_lock.

ahrens commented:
@behlendorf I've updated to code and I think this addresses all the issues that have been raised.  

@bzzz77 could you re-evaluate the performance of dnode_setdirty()?

behlendorf commented:
@ahrens great I'll go over it again.  I've asked @tonyhutter to give #5764 a review too so can get it merged, then this PR can be squashed and rebased on those changes.

ahrens commented:
@behlendorf FYI, this is already rebased on #5764, it's the first 4 commits.  But yeah we can squash to make it more clear.

behlendorf commented:
OK, #5764 was squashed and merged to master.  If you can rebase and squash this it'll make it easier to review.  Thanks.

ahrens commented:
@behlendorf Done.

bzzz77 commented:
I can confirm that the last version of dnode_setdirty() is doing great - I can achieve rougly the same numbers as with my patch.

behlendorf commented on module/zfs/dsl_dataset.c:
```diff
@@ -1704,6 +1704,9 @@ dsl_dataset_sync_done(dsl_dataset_t *ds, dmu_tx_t *tx)
 	bplist_iterate(&ds->ds_pending_deadlist,
 	    deadlist_enqueue_cb, &ds->ds_deadlist, tx);
 
+	if (dmu_objset_userused_enabled(os))
```
The ASSERTV above needs to be removed.

dinatale2 commented:
I recommend you rebase on master to get rid of the style errors. Those were corrected yesterday.

ahrens commented:
@behlendorf this has been pushed to OpenZFS and illumos https://github.com/illumos/illumos-gate/commit/94c2d0eb22e9624151ee84a7edbf7178e1bf4087

behlendorf commented:
@ahrens great!  Could you please squash and rebase this one last time so we can get a final test run through the bots.

ahrens commented:
@behlendorf done.

behlendorf commented:
Thanks!
